### ISSUE TITLE
GitCommands.FullPathResolver.Resolve throws System.UriFormatException

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#6235]
## Current behaviour

`Resolve("drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c")` results in `fullPath = "\\\\.\\aux"` which caues `new Uri(fullPath)` to throw an `UriFormatException` (while playing around with the Linux repo).

It seems that `System.IO.Path.GetFullPath()` handles file names which contain reserved DOS device names like NUL, CON, AUX, PRN, LST, COM1, LPT1, etc., in an unexpected manner.
Windows does not allow to create files named "aux.*" with any extension.

## Expected behaviour

no unhandled exception

## Steps to reproduce

open the Commit dialog for the Linux repo with the source file `drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c` removed

## Did this work in previous version of GitExtensions

yes, 2.51.05
no, 3.0.0

## Environment

- Git Extensions 3.1.0
- Build aa2e8fc3e20a340b9a1e366dde127dcb537ff9e4
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.2117.0
- DPI 96dpi (no scaling)

## Diagnostics

```
  at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at GitCommands.FullPathResolver.Resolve(String path) in F:\Build\gitextensions\GitCommands\FullPathResolver.cs:line 54
   at GitUI.Editor.FileViewer.<ShowOrDeferAsync>g__GetFileLength|82_0(<>c__DisplayClass82_0& ) in F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 412
   at GitUI.Editor.FileViewer.ViewCurrentChanges(String fileName, String oldFileName, Boolean staged, Boolean isSubmodule, Func`1 getStatusAsync, Action openWithDifftool) in F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 470
   at GitUI.Editor.FileViewer.ViewCurrentChanges(GitItemStatus item, Boolean isStaged, Action openWithDifftool) in F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 463
   at GitUI.CommandsDialogs.FormCommit.<SetSelectedDiffAsync>d__126.MoveNext() in F:\Build\gitextensions\GitUI\CommandsDialogs\FormCommit.cs:line 1173
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.CommandsDialogs.FormCommit.<>c__DisplayClass125_0.<<ShowChanges>b__0>d.MoveNext() in F:\Build\gitextensions\GitUI\CommandsDialogs\FormCommit.cs:line 1153
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext() in F:\Build\gitextensions\GitExtUtils\GitUI\ThreadHelper.cs:line 101
```

```
2019-02-10T18:11:56.0107553+01:00	77	13640	UI	1	git	config --get ą	
2019-02-10T18:11:56.0917656+01:00	79	10696	UI	1	git		
2019-02-10T18:11:57.3269224+01:00	66	20572	UI	0	git	--version	
2019-02-10T18:11:57.4109331+01:00	154	25144		0	git	--no-optional-locks for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	F:\Build\Linux\
2019-02-10T18:11:57.4224346+01:00	200	12932		0	git	--no-optional-locks show-ref --dereference	F:\Build\Linux\
2019-02-10T18:11:57.4964440+01:00	133	10404	UI	0	git	rev-parse --git-common-dir	F:\Build\Linux\
2019-02-10T18:11:57.5844551+01:00	98	11320		0	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/	F:\Build\Linux\
2019-02-10T18:11:57.6314611+01:00	102	6536		0	git	remote -v	F:\Build\Linux\
2019-02-10T18:11:57.7454756+01:00	91	11028		0	git	show-ref --dereference	F:\Build\Linux\
2019-02-10T18:11:57.8429880+01:00	78	11236		0	git	remote	F:\Build\Linux\
2019-02-10T18:11:58.1010207+01:00	91	16952		0	git	for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	F:\Build\Linux\
2019-02-10T18:11:58.1650289+01:00	97	18536		0	git	ls-files -z --unmerged	F:\Build\Linux\
2019-02-10T18:11:58.2245364+01:00	77	13484	UI	0	git	rev-parse HEAD	F:\Build\Linux\
2019-02-10T18:11:58.3115475+01:00	98	14620	UI	0	git	rev-parse HEAD	F:\Build\Linux\
2019-02-10T18:11:58.4220615+01:00	117	26012	UI	0	git	show-ref --dereference	F:\Build\Linux\
2019-02-10T18:11:58.5970837+01:00	113	21624		0	git	for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	F:\Build\Linux\
2019-02-10T18:11:58.6170863+01:00	5337	8268			git	-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b" --all --boundary --max-count=100000 --	F:\Build\Linux\
2019-02-10T18:11:58.6370888+01:00	129	12804	UI	0	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/master	F:\Build\Linux\
2019-02-10T18:11:58.8021098+01:00	99	21632		0	git	ls-files -z --unmerged	F:\Build\Linux\
2019-02-10T18:11:58.8396145+01:00	94	23132	UI	0	git	rev-parse "master~0"	F:\Build\Linux\
2019-02-10T18:11:58.9826327+01:00	321	8640		0	git	--no-optional-locks status --porcelain=2 -z --untracked-files --ignore-submodules=none	F:\Build\Linux\
2019-02-10T18:11:59.6012112+01:00	117	15376	UI	0	git	-c diff.submodule=short -c diff.noprefix=false diff --no-color -M -C -z --name-status "f014ffb025c159fd51d19af8af0022a991aaa4f8" "bab5c80b211035739997ebd361a679fa85b39465"	F:\Build\Linux\
2019-02-10T18:11:59.7277273+01:00	152	22960	UI	0	git	-c diff.submodule=short -c diff.noprefix=false diff --no-color -M -C -z --name-status "3f4258bbe0363b3d5330b93de0e86b7796ff1a74" "bab5c80b211035739997ebd361a679fa85b39465"	F:\Build\Linux\
2019-02-10T18:11:59.8842472+01:00	184	19760	UI	0	git	diff-tree --name-only -z --cc --no-commit-id bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.3053006+01:00	100	9248	UI	0	git	-c log.showSignature=false log -1 --pretty="format:%H%n%e%n%B%nNotes:%n%-N" bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.4313166+01:00	152	5560		0	git	for-each-ref --sort=-committerdate --sort=-taggerdate --format="%(refname)" refs/	F:\Build\Linux\
2019-02-10T18:12:00.4403178+01:00	248	11396		0	git	branch -a --contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.4458185+01:00	238	14928		0	git	tag --contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.4463185+01:00	325	11036	UI	0	git	-c log.showSignature=false log -1 --pretty="format:%H%n%e%n%B%nNotes:%n%-N" bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.4648209+01:00	284	23216		0	git	describe --tags --first-parent --abbrev=40 bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.7848615+01:00	133	8112		0	git	describe --tags --first-parent --abbrev=40 bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.7848615+01:00	192	7012		0	git	tag --contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.7848615+01:00	196	20824		0	git	branch -a --contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\
2019-02-10T18:12:00.9223790+01:00	99	13412		0	git	-c diff.submodule=short -c diff.noprefix=false diff --no-color --unified=3 -M -C "f014ffb025c159fd51d19af8af0022a991aaa4f8" "bab5c80b211035739997ebd361a679fa85b39465" -- "arch/arm/boot/dts/imx53-qsb-common.dtsi"	F:\Build\Linux\
2019-02-10T18:12:01.9150050+01:00	82	25088	UI	128	git	rev-parse "MERGE_HEAD~0"	F:\Build\Linux\
2019-02-10T18:12:01.9205057+01:00	335	19264		0	git	status --porcelain=2 -z --untracked-files --ignore-submodules=none	F:\Build\Linux\
2019-02-10T18:12:02.2765509+01:00	78	8368	UI	0	git	rev-parse "HEAD~0"	F:\Build\Linux\
2019-02-10T18:12:02.3595615+01:00	93	24644	UI	0	git	ls-files -z --unmerged	F:\Build\Linux\
2019-02-10T18:12:40.9064563+01:00	110	23444		0	git	ls-files -z --unmerged	F:\Build\Linux\
```

**Comment by RussKie** [comment]
IIRC we have at least one open issue about this

On Feb 11, 2019 7:26 AM, "Michael Seibt" <notifications@github.com> wrote:

Current behaviour

Resolve("drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c") results in fullPath
= "\\\\.\\aux" which caues new Uri(fullPath) to throw an UriFormatException
(while playing around with the Linux repo).

It seems that System.IO.Path.GetFullPath() handles file names which contain
reserved DOS device names like NUL, CON, AUX, PRN, LST, COM1, LPT1, etc.,
in an unexpected manner.
Windows does not allow to create files named "aux.*" with any extension.
Expected behaviour

no unhandled exception
Steps to reproduce

open the Commit dialog for the Linux repo with the source file
drivers/gpu/drm/nouveau/nvkm/subdev/i2c/aux.c modified
Did this work in previous version of GitExtensions

yes, 2.51.05
no, 3.0.0
Environment

   - Git Extensions 3.1.0
   - Build aa2e8fc
   <https://github.com/gitextensions/gitextensions/commit/aa2e8fc3e20a340b9a1e366dde127dcb537ff9e4>
   - Git 2.20.1.windows.1
   - Microsoft Windows NT 6.1.7601 Service Pack 1
   - .NET Framework 4.7.2117.0
   - DPI 96dpi (no scaling)

Diagnostics

  at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)

   at GitCommands.FullPathResolver.Resolve(String path) in
F:\Build\gitextensions\GitCommands\FullPathResolver.cs:line 54

   at GitUI.Editor.FileViewer.<ShowOrDeferAsync>g__GetFileLength|82_0(<>c__DisplayClass82_0&
) in F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 412

   at GitUI.Editor.FileViewer.ViewCurrentChanges(String fileName,
String oldFileName, Boolean staged, Boolean isSubmodule, Func`1
getStatusAsync, Action openWithDifftool) in
F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 470

   at GitUI.Editor.FileViewer.ViewCurrentChanges(GitItemStatus item,
Boolean isStaged, Action openWithDifftool) in
F:\Build\gitextensions\GitUI\Editor\FileViewer.cs:line 463

   at GitUI.CommandsDialogs.FormCommit.<SetSelectedDiffAsync>d__126.MoveNext()
in F:\Build\gitextensions\GitUI\CommandsDialogs\FormCommit.cs:line
1173

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task
task)

   at GitUI.CommandsDialogs.FormCommit.<>c__DisplayClass125_0.<<ShowChanges>b__0>d.MoveNext()
in F:\Build\gitextensions\GitUI\CommandsDialogs\FormCommit.cs:line
1153

--- End of stack trace from previous location where exception was thrown ---

   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task
task)

   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
in F:\Build\gitextensions\GitExtUtils\GitUI\ThreadHelper.cs:line 101


2019-02-10T18:11:56.0107553+01:00	77	13640	UI	1	git	config --get ą	

2019-02-10T18:11:56.0917656+01:00	79	10696	UI	1	git		

2019-02-10T18:11:57.3269224+01:00	66	20572	UI	0	git	--version	

2019-02-10T18:11:57.4109331+01:00	154	25144		0	git	--no-optional-locks
for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname)
%(refname)"	F:\Build\Linux\

2019-02-10T18:11:57.4224346+01:00	200	12932		0	git	--no-optional-locks
show-ref --dereference	F:\Build\Linux\

2019-02-10T18:11:57.4964440+01:00	133	10404	UI	0	git	rev-parse
--git-common-dir	F:\Build\Linux\

2019-02-10T18:11:57.5844551+01:00	98	11320		0	git	for-each-ref
--format="%(push:track)::%(upstream:track)::%(refname:short)"
refs/heads/	F:\Build\Linux\

2019-02-10T18:11:57.6314611+01:00	102	6536		0	git	remote -v	F:\Build\Linux\

2019-02-10T18:11:57.7454756+01:00	91	11028		0	git	show-ref
--dereference	F:\Build\Linux\

2019-02-10T18:11:57.8429880+01:00	78	11236		0	git	remote	F:\Build\Linux\

2019-02-10T18:11:58.1010207+01:00	91	16952		0	git	for-each-ref
--sort=-committerdate refs/heads/ --format="%(objectname)
%(refname)"	F:\Build\Linux\

2019-02-10T18:11:58.1650289+01:00	97	18536		0	git	ls-files -z
--unmerged	F:\Build\Linux\

2019-02-10T18:11:58.2245364+01:00	77	13484	UI	0	git	rev-parse
HEAD	F:\Build\Linux\

2019-02-10T18:11:58.3115475+01:00	98	14620	UI	0	git	rev-parse
HEAD	F:\Build\Linux\

2019-02-10T18:11:58.4220615+01:00	117	26012	UI	0	git	show-ref
--dereference	F:\Build\Linux\

2019-02-10T18:11:58.5970837+01:00	113	21624		0	git	for-each-ref
--sort=-committerdate refs/heads/ --format="%(objectname)
%(refname)"	F:\Build\Linux\

2019-02-10T18:11:58.6170863+01:00	5337	8268			git	-c
log.showSignature=false log -z
--pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b"
--all --boundary --max-count=100000 --	F:\Build\Linux\

2019-02-10T18:11:58.6370888+01:00	129	12804	UI	0	git	for-each-ref
--format="%(push:track)::%(upstream:track)::%(refname:short)"
refs/heads/master	F:\Build\Linux\

2019-02-10T18:11:58.8021098+01:00	99	21632		0	git	ls-files -z
--unmerged	F:\Build\Linux\

2019-02-10T18:11:58.8396145+01:00	94	23132	UI	0	git	rev-parse
"master~0"	F:\Build\Linux\

2019-02-10T18:11:58.9826327+01:00	321	8640		0	git	--no-optional-locks
status --porcelain=2 -z --untracked-files
--ignore-submodules=none	F:\Build\Linux\

2019-02-10T18:11:59.6012112+01:00	117	15376	UI	0	git	-c
diff.submodule=short -c diff.noprefix=false diff --no-color -M -C -z
--name-status "f014ffb025c159fd51d19af8af0022a991aaa4f8"
"bab5c80b211035739997ebd361a679fa85b39465"	F:\Build\Linux\

2019-02-10T18:11:59.7277273+01:00	152	22960	UI	0	git	-c
diff.submodule=short -c diff.noprefix=false diff --no-color -M -C -z
--name-status "3f4258bbe0363b3d5330b93de0e86b7796ff1a74"
"bab5c80b211035739997ebd361a679fa85b39465"	F:\Build\Linux\

2019-02-10T18:11:59.8842472+01:00	184	19760	UI	0	git	diff-tree
--name-only -z --cc --no-commit-id
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.3053006+01:00	100	9248	UI	0	git	-c
log.showSignature=false log -1
--pretty="format:%H%n%e%n%B%nNotes:%n%-N"
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.4313166+01:00	152	5560		0	git	for-each-ref
--sort=-committerdate --sort=-taggerdate --format="%(refname)"
refs/	F:\Build\Linux\

2019-02-10T18:12:00.4403178+01:00	248	11396		0	git	branch -a
--contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.4458185+01:00	238	14928		0	git	tag --contains
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.4463185+01:00	325	11036	UI	0	git	-c
log.showSignature=false log -1
--pretty="format:%H%n%e%n%B%nNotes:%n%-N"
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.4648209+01:00	284	23216		0	git	describe --tags
--first-parent --abbrev=40
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.7848615+01:00	133	8112		0	git	describe --tags
--first-parent --abbrev=40
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.7848615+01:00	192	7012		0	git	tag --contains
bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.7848615+01:00	196	20824		0	git	branch -a
--contains bab5c80b211035739997ebd361a679fa85b39465	F:\Build\Linux\

2019-02-10T18:12:00.9223790+01:00	99	13412		0	git	-c
diff.submodule=short -c diff.noprefix=false diff --no-color
--unified=3 -M -C "f014ffb025c159fd51d19af8af0022a991aaa4f8"
"bab5c80b211035739997ebd361a679fa85b39465" --
"arch/arm/boot/dts/imx53-qsb-common.dtsi"	F:\Build\Linux\

2019-02-10T18:12:01.9150050+01:00	82	25088	UI	128	git	rev-parse
"MERGE_HEAD~0"	F:\Build\Linux\

2019-02-10T18:12:01.9205057+01:00	335	19264		0	git	status
--porcelain=2 -z --untracked-files
--ignore-submodules=none	F:\Build\Linux\

2019-02-10T18:12:02.2765509+01:00	78	8368	UI	0	git	rev-parse
"HEAD~0"	F:\Build\Linux\

2019-02-10T18:12:02.3595615+01:00	93	24644	UI	0	git	ls-files -z
--unmerged	F:\Build\Linux\

2019-02-10T18:12:40.9064563+01:00	110	23444		0	git	ls-files -z
--unmerged	F:\Build\Linux\


—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/gitextensions/gitextensions/issues/6235>, or mute the
thread

**Comment by gerhardol** [comment]
See #5956 for a PR by @sharwell not merged as the test fails

**Comment by mstv** [comment]
Hard to find without a bug report.

**Comment by mstv** [comment]
"Fixes #6235" should be added to the PR #5956.

**Comment by RussKie** [comment]
done
